packages/client/src/queries/applyQuery.ts

```diff
@@ -0,0 +1,330 @@
+  client: MinimalClient,
+  query: QD,
+  params?: P,
+): Promise<
+  QueryReturnType<QD["output"]>
+> {
+  const response = await OntologiesV2.QueryTypes.executeQueryV2(
+    addUserAgent(client, query),
+    client.ontologyRid,
+    query.apiName,
+    {
+      parameters: params
+        ? await remapQueryParams(
+          params as { [parameterId: string]: any },
+          client,
+          query.parameters,
+        )
+        : {},
+    },
+  );
+  const objectOutputDefs = await getRequiredDefinitions(query.output, client);
+  const remappedResponse = await remapQueryResponse(
+    client,
+    query.output,
+    response.value,
+    objectOutputDefs,
+  );
+  return remappedResponse as QueryReturnType<QD["output"]>;
+}
+
+async function remapQueryParams(
+  params: { [parameterId: string]: any },
+  client: MinimalClient,
+  paramTypes: Record<string, QueryParameterDefinition<any>>,
+): Promise<{ [parameterId: string]: any }> {
+  const parameterMap: { [parameterName: string]: unknown } = {};
+  for (const [key, value] of Object.entries(params)) {
+    parameterMap[key] = await toDataValueQueries(
+      value,
+      client,
+      paramTypes[key],
+    );
+  }
+  return parameterMap;
+}
+
+async function remapQueryResponse<
+  K extends string,
+  Q extends ObjectTypeDefinition<any>,
+  T extends QueryDataTypeDefinition<K, Q | never>,
+>(
+  client: MinimalClient,
+  responseDataType: T,
+  responseValue: DataValue,
+  definitions: Map<string, ObjectOrInterfaceDefinition>,
+): Promise<QueryReturnType<T>> {
+  // handle null responses
+  if (responseValue == null) {
+    if (responseDataType.nullable) {
+      return undefined as unknown as QueryReturnType<typeof responseDataType>;
+    } else {
+      throw new Error("Got null response when nullable was not allowed");
+    }
+  }
+
+  if (
+    responseDataType.multiplicity != null
+    && responseDataType.multiplicity !== false
+  ) {
+    const withoutMultiplicity = { ...responseDataType, multiplicity: false };
+    for (let i = 0; i < responseValue.length; i++) {
+      responseValue[i] = remapQueryResponse(
+        responseValue[i],
+        withoutMultiplicity,
+        client,
+        definitions,
+      );
+    }
+    return responseValue as QueryReturnType<typeof responseDataType>;
+  }
+
+  switch (responseDataType.type) {
+    case "union": {
+      throw new Error("Union return types are not yet supported");
+    }
+
+    case "set": {
+      for (let i = 0; i < responseValue.length; i++) {
+        responseValue[i] = remapQueryResponse(
+          responseValue[i],
+          responseDataType.set,
+          client,
+          definitions,
+        );
+      }
+
+      return responseValue as QueryReturnType<typeof responseDataType>;
+    }
+
+    case "attachment": {
+      return createAttachmentFromRid(client, responseValue) as QueryReturnType<
+        typeof responseDataType
+      >;
+    }
+    case "object": {
+      const def = responseDataType.__OsdkTargetType
```

Comment:
So if for some reason the objectDef wasn't capture in the type, I also just try manually loading the def from the `ontologyProvider` and then throw if both times I can't find it. Not sure if this is overkill though...

Reply:
Changed to just use provider